scripts/uploadBundles.ts

```diff
@@ -0,0 +1,175 @@
+import {readdir, readFile, stat, writeFile} from 'node:fs/promises'
+import path from 'node:path'
+
+// eslint-disable-next-line import/no-extraneous-dependencies
+import {Storage, type UploadOptions} from '@google-cloud/storage'
+
+import {readEnv} from './utils/envVars'
+
+type KnownEnvVar = 'GOOGLE_PROJECT_ID' | 'GCLOUD_SERVICE_KEY' | 'GCLOUD_BUCKET'
+
+const storage = new Storage({
+  projectId: readEnv<KnownEnvVar>('GOOGLE_PROJECT_ID'),
+  credentials: JSON.parse(readEnv<KnownEnvVar>('GCLOUD_SERVICE_KEY')),
+})
+
+const bucket = storage.bucket(readEnv<KnownEnvVar>('GCLOUD_BUCKET'))
+
+const corePkgs = ['sanity', '@sanity/vision'] as const
+const sharedPkgs = ['react', 'react-dom', 'styled-components'] as const
+
+const appVersion = 'v1'
+
+/**
+ * Recursively iterate through a directory and yield all files
+ * @param directory - The directory to iterate
+ */
+async function* getFiles(directory: string): AsyncGenerator<string, void, unknown> {
+  for (const file of await readdir(directory)) {
+    const fullPath = path.join(directory, file)
+    const stats = await stat(fullPath)
+
+    if (stats.isDirectory()) {
+      yield* getFiles(fullPath)
+    }
+
+    if (stats.isFile()) {
+      yield fullPath
+    }
+  }
+}
+
+async function copyPackages() {
+  console.log('**Copying Core packages**')
+
+  const packageVersions = new Map<string, string>()
+
+  // First we iterate through each core package located in `packages/`
+  for (const pkg of corePkgs) {
+    console.log(`Copying files from ${pkg}`)
+
+    const packageJson = JSON.parse(await readFile(`packages/${pkg}/package.json`, 'utf8'))
+
+    const {version} = packageJson
+    packageVersions.set(pkg, version)
+
+    // Convert slashes to double underscores
+    // Needed for `@sanity/vision`
+    const cleanDir = pkg.replace(/\//g, '__')
+
+    for await (const filePath of getFiles(`packages/${pkg}/dist`)) {
+      try {
+        const fileName = path.basename(filePath)
+        const options: UploadOptions = {
+          destination: `modules/${appVersion}/${cleanDir}/${version}/bare/${fileName}`,
+          gzip: true,
+          contentType: 'application/javascript',
+        }
+
+        // Upload files to the proper bucket destination
+        await bucket.upload(filePath, options)
```

Comment:
I may be mistaken, but it looks like files will be uploaded serially. Would it be worthwhile to parallelise this work?

Reply:
(Same for the shared packages).